### DISCUSSION TITLE
Development questions

### DISCUSSION THREAD

**Discussion by mickeprag** [kizniche/Mycodo#1406] [Q&A]
Hi,

I am working on integrating some new hardware into Mycodo and have some questions.

1. Is there planned to support async functions in plugin code?
2. Do input and output plugins need two separate plugins? Or can one plugin define both inputs and outputs?
  2.1. If the answer is two separate, where is the best place to store code common for both types of plugins?
3. Does the `measurements_dict`/`channels_dict` has to be known in advance or can it be populated at runtime? If so, how?
4. I want to populate a dropdown at runtime. I can do this for an input plugin since the code is executed before the settings dialog is opened. For an output plugin it will be executed as soon as the plugin is added and will be too late. It might be hard to explain...

Thanks

**Comment by kizniche** [comment]
Hi. I'll try to answer all I can.

1. Yes, this already can be used, if I understand you correctly. Asyncio is already in use in at least one input module.
2. No, you can combine input and output functionality in one module, though it does help system integration if you do separate these functionalities, since other parts of the system could benefit from accessing these outputs. If the output functionality was integrated into an input module, the output would only be able to use used by that input module.
3. Generally, yes, it needs to be defined when the Python file is loaded. What use case do you have that there will be a variable number of measurements? There are some Input modules that operate with a variable number of measurements (such as MQTT and TTN), which rely on a user input value to determine how many measurements there are, and these can be used as a reference.
4. All inputs need to be defined at the load of the Python file. If you explain the purpose or desired behavior, I can tell you if it can be done.

**Comment by mickeprag** [comment]
1. Do you know which one?
2. If (when) I push the code upstream, this is not an issue. But before then, using the module will require the user to install two separate plugins. Also sharing code between them will be harder. Not a bit issue, just an inconvenience...
3. The use case is that I am trying to integrate a hardware that will be able to tell Mycodo what kind of inputs/outputs it supports. But this interview will happen at runtime. If the measurements need to be static then I need to write one input module for every hardware variant, else I could create one generic and depending on what type of board you connect, the correct inputs/outputs would be presented to the user.
I can see that the mqtt module has an empty `measurements_dict = {}`. I will try to look at that code to see if I can understand how it works...
4. I have a CAN bus where any number of boards may be connected. These will be discovered automatically by my code and I want to populate a dropdown where the user may choose which board to use. This is working for the input module but not the output module.
The reason this works for the input module is that the input module is initially added as deactivated. During this time my code will be imported and the discovery routine can take place. Once the user opens the settings dialog the Python code knows which boards are available and can show the boards available.
<img width="455" alt="image" src="https://github.com/user-attachments/assets/d3ae7e1b-3143-4f3e-b982-4cd3248b0fa7" />
<img width="455" alt="image" src="https://github.com/user-attachments/assets/643062da-43ad-4856-bd38-837212be84f3" />
